lib/cucumber/filters/retry.rb

```diff
@@ -0,0 +1,32 @@
+require 'cucumber/core/filter'
+require 'cucumber/running_test_case'
+require 'cucumber/events/bus'
+require 'cucumber/events/after_test_case'
+
+module Cucumber
+  module Filters
+    class Retry < Core::Filter.new(:configuration)
+
+      def test_case(test_case)
+        configuration.on_event(:after_test_case) do |event|
+          next unless retry_required?(test_case, event)
+
+          test_case_counts[test_case] += 1
+          event.test_case.describe_to(receiver)
+        end
+
+        super
+      end
+
+      private
+
+      def retry_required?(test_case, event)
+        event.test_case == test_case && event.result.failed? && test_case_counts[test_case] < configuration.retry_attempts
```

Comment:
I've resolved this with https://github.com/cucumber/cucumber-ruby-core/pull/111 which makes equality for `Test::Case` objects work as you would expect it to.